src/polynomial/mod.rs

```diff
@@ -244,15 +304,17 @@ impl_expression_ops!(Sub, sub, Sum, Expression<F>, Neg::neg);
 impl_expression_ops!(Mul, mul, Product, Expression<F>, std::convert::identity);
 
 /// Monomial: c_i*x_0^d_0*..*x_{n-1}^d_{n-1} with arity n
-/// index_to_poly is mapping from i to (rotation, column_index)
-/// poly_to_index is mapping from (rotation, column_index) to i \in [0,..,n-1]
+/// index_to_poly is mapping from i to (rotation, column_index, type)
+/// poly_to_index is mapping from (rotation, column_index, type) to i \in [0,..,n-1]
+/// type has value either CHALLENGE_TYPE or POLYNOMIAL_TYPE
 #[derive(Clone)]
 pub struct Monomial<F: PrimeField> {
     pub arity: usize,
-    // poly: (rotation, column_index)
-    pub index_to_poly: Vec<(i32, usize)>,
-    // (rotation, column_index) -> index
-    pub poly_to_index: HashMap<(i32, usize), usize>,
+    // poly or challenge: (rotation, column_index, type)
+    pub index_to_poly: Vec<(i32, usize, usize)>,
+    // (rotation, column_index, var_type) -> index
+    // when type = CHALLENGE_TYPE as a challenge, we always set rotation = 0
+    pub poly_to_index: HashMap<(i32, usize, usize), usize>,
```

Comment:
So as not to delay the process, I will allocate a separate structure myself later, instead of tuple in so many places